### PR TITLE
[baseline][opengl,mesa] Resolve ambiguities

### LINKED ISSUE #28087: `#include <GL/gl.h>` is ambiguous on case-insensitve file systems
**Describe the bug**

osg:x64-windows-static-md currently fails in CI during the build step because `GLint64` and `GLuint64` appear as undefined:
~~~
D:\buildtrees\osg\src\raph-3.6.5-41fa3cfd98.clean\include\osg/GLExtensions(323): error C2061: syntax error: identifier 'GLint64'
D:\buildtrees\osg\src\raph-3.6.5-41fa3cfd98.clean\include\osg/GLExtensions(325): error C2061: syntax error: identifier 'GLuint64'
~~~
But during the configure step, the types are found:
~~~
-- Found OpenGL: opengl32   
-- Found EGL: D:/installed/x64-windows-static-md/debug/lib/libEGL.lib  
-- Performing Test GL_HEADER_HAS_GLINT64
-- Performing Test GL_HEADER_HAS_GLINT64 - Success
-- Performing Test GL_HEADER_HAS_GLUINT64
-- Performing Test GL_HEADER_HAS_GLUINT64 - Success
~~~
with the following header:
~~~
//#include<> line for OpenGL Header
OPENGL_HEADER1:STRING=#include <GL/gl.h>

//#include<> line for additional OpenGL Headers if required
OPENGL_HEADER2:STRING=
~~~

Further investigation showed that both ports opengl and mesa provide this header, only differing in case, from the top-level dir:
~~~
mesa:x64-windows-static-md:/include/GL/gl.h
opengl:x64-windows-static-md:/include/gl/GL.h
~~~
I have no idea if they are identical or differ.

FTR additional matches do exist in subdirs, possibly becoming available via the include path:
~~~
pangolin:x64-windows-static-md:/include/pangolin/gl/gl.h
glbinding:x64-windows-static-md:/include/glbinding/gl/gl.h
gstreamer:x64-windows-static-md:/include/gstreamer-1.0/gst/gl/gl.h
magnum:x64-windows-static-md:/include/Magnum/GL/GL.h
~~~~

**Environment**
vcpkg CI, x64-mingw-static-md

**Failure logs**
https://dev.azure.com/vcpkg/public/_build/results?buildId=81637&view=artifacts&pathAsName=false&type=publishedArtifacts

CC @Neumann-A

### LINKED ISSUE #28688: [egl] Build failure
**Host Environment**

- OS: Windows 11 
- Compiler: MinGW-w64 [MSYS2] 
- Triplet: x64-mingw-dynamic

**To Reproduce**

`./vcpkg install egl`

**Failure logs**

- Build message:
```
Computing installation plan...
The following packages will be built and installed:
    egl[core]:x64-mingw-dynamic -> 2022-12-04
Detecting compiler hash for triplet x64-mingw-dynamic...
Restored 0 package(s) from C:\Users\buele\AppData\Local\vcpkg\archives in 170 us. Use --debug to see more details.
Installing 1/1 egl:x64-mingw-dynamic...
Building egl[core]:x64-mingw-dynamic...
warning: -- Using community triplet x64-mingw-dynamic. This triplet configuration is not guaranteed to succeed.
-- [COMMUNITY] Loading triplet configuration from: C:\DevTools\vcpkg\triplets\community\x64-mingw-dynamic.cmake
-- Fixing pkgconfig file: C:/DevTools/vcpkg/packages/egl_x64-mingw-dynamic/lib/pkgconfig/egl.pc
-- Using cached msys-mingw-w64-i686-pkg-config-0.29.2-3-any.pkg.tar.zst.
-- Using cached msys-mingw-w64-i686-libwinpthread-git-9.0.0.6373.5be8fcd83-1-any.pkg.tar.zst.
-- Using msys root at C:/DevTools/vcpkg/downloads/tools/msys2/9a1ec3f33446b195
CMake Error at scripts/cmake/vcpkg_fixup_pkgconfig.cmake:31 (message):

  C:/DevTools/vcpkg/downloads/tools/msys2/9a1ec3f33446b195/mingw32/bin/pkg-config.exe
  --exists egl failed with error code: 1

      ENV{PKG_CONFIG_PATH}: "C:/DevTools/vcpkg/packages/egl_x64-mingw-dynamic/lib/pkgconfig;C:/DevTools/vcpkg/packages/egl_x64-mingw-dynamic/share/pkgconfig;C:/DevTools/vcpkg/installed/x64-mingw-dynamic/lib/pkgconfig;C:/DevTools/vcpkg/installed/x64-mingw-dynamic/share/pkgconfig"
      output: Package opengl was not found in the pkg-config search path.

  Perhaps you should add the directory containing `opengl.pc'

  to the PKG_CONFIG_PATH environment variable

  Package 'opengl', required by 'egl', not found
Call Stack (most recent call first):
  scripts/cmake/vcpkg_fixup_pkgconfig.cmake:154 (z_vcpkg_fixup_pkgconfig_check_files)
  ports/egl/portfile.cmake:10 (vcpkg_fixup_pkgconfig)
  scripts/ports.cmake:147 (include)
  ```
- buildtrees/egl/x64-mingw-dynamic.vcpkg_abi_info.txt
``` 
ENV:PATH 629d3e7738a2270aa0caf4827eb9a3a1385b176ed8e4390d8280ffab68922b68
angle 1d6aa9980a7708c3afcaa4847004c85985e105f8f32c6f8eed6add79c14d1e19
cmake 3.25.1
egl.pc.in 4f9ea65045b991d2a301181741f6ade7b37a6aa8a57198f80f2dffe70af4aad9
features core
opengl 839f27146a6f4b0da9345e48d05d2c3db0e5a4b7709f6c41aedac0ba00706d9d
portfile.cmake 71834259c4a3251a020f1c80bb06c125a3fed495145cc8ff022b532a7f374d1a
ports.cmake b4accc7941cb5ff993e1a0434cea6df3e99331137294958437b2b86b7216e2e2
post_build_checks 2
powershell 7.3.1
triplet x64-mingw-dynamic
triplet_abi 621e918dca7fdc6cb403bce0ba31dcf1fc68dfe47bec975bdcc66f0ea79dc548-91dd9574fdb1f05fff793a5f39126d3404f6bf303be55a2e7f81b71a14396041-7ad91d8e1752a5cc1e5aa4d57fd66a89629a04bb
vcpkg-cmake-wrapper.cmake bfe7ff2ffa0884ad457a11f4ddfbc04d824bb17b884e0c3e228b5ef5cfb79218
vcpkg.json 00125c0ec775964828fb1e18affb5bf04fa79ef5f2d2e261ad3ba93b77fa4630
vcpkg_fixup_pkgconfig 433d0d235b4e9f6c7bb8744a2f06a9f534b355da67d0d85b7c2daf15c8c77671
```

**Additional context**

* `git pull` and `vcpkg update`
* looking for `opengl.pc` (no file found)

### PR DESCRIPTION
**Pull request by dg0yt**
- #### What does your PR fix?
  Fix a CI baseline regression for x64-windows-static-md, port osg.
  Fixes #28087, by ~no longer copying libs and headers out of the Windows SDK at all (ultimately) /~ using wrong case and by changes to mesa.

  I fail to see a *robust* benefit from the copying of SDK headers. AFAIU package opengl's ABI hash already doesn't depend on the Windows SDK, i.e. rebuilds of the port will be triggered independently of changes to the active version of the Windows SDK, picking what they will find at this time. And even just using exported artifacts on another machine should demand using a compatible (i.e. controlled) version of the SDK.

  Somewhat related, this PR moves the opengl32 lib generated by mesa to manual-link, and drops all mesa headers which are available from opengl-registry.

  Fixes  #28688

  Revise mesa dependencies.
  Installs a proper copyright file for mesa.

  Drafts of these changes were tested and discussed in https://github.com/microsoft/vcpkg/pull/28089.

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?
  unchanged, yes

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
  yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?
  yes

### CODE REVIEW

File: ports/mesa/portfile.cmake [hunk 1]
```diff
@@ -117,24 +117,33 @@ vcpkg_configure_meson(
...
+endif()
+
+if(VCPKG_TARGET_IS_WINDOWS AND EXISTS "${CURRENT_PACKAGES_DIR}/lib/opengl32${VCPKG_TARGET_IMPORT_LIBRARY_SUFFIX}")
+    # Avoid collision with opengl32 lib from Windows SDK (or mingw).
+    file(MAKE_DIRECTORY "${CURRENT_PACKAGES_DIR}/lib/manual-link")
+    file(RENAME "${CURRENT_PACKAGES_DIR}/lib/opengl32${VCPKG_TARGET_IMPORT_LIBRARY_SUFFIX}" "${CURRENT_PACKAGES_DIR}/lib/manual-link/opengl32${VCPKG_TARGET_IMPORT_LIBRARY_SUFFIX}")
+    if(NOT VCPKG_BUILD_TYPE)
+        file(MAKE_DIRECTORY "${CURRENT_PACKAGES_DIR}/debug/lib/manual-link")
+        file(RENAME "${CURRENT_PACKAGES_DIR}/debug/lib/opengl32${VCPKG_TARGET_IMPORT_LIBRARY_SUFFIX}" "${CURRENT_PACKAGES_DIR}/debug/lib/manual-link/opengl32${VCPKG_TARGET_IMPORT_LIBRARY_SUFFIX}")
+    endif()
```
Comment (Neumann-A) [inline ports/mesa/portfile.cmake:146]: I don't think this is necessary. The way this works is just to make the linker happy about the opengl symbols. 
Which DLL actually gets loaded depends on load order. the opengl32.dll provided by mesa is a drop in replacement providing full opengl software rendering.

**Comment by mrowrpurr** [comment]
Friendly ping. This is blocking #28126

File: ports/mesa/portfile.cmake (hunk 1 above)
Comment (BillyONeal) [reply ports/mesa/portfile.cmake:146]: If the symbols exported by this lib are identical to the Windows SDK ones, the user should get the Windows SDK ones (which we require to be present anyway) and the ones in manual-link are pointless (the Windows SDK ones will always be found first).

If the set of symbols differs we need to come up with something else...

Comment (dg0yt) [reply ports/mesa/portfile.cmake:146]: I cannot check the Windows SDK files. But AFAIU mesa explicitly claims to provide a drop-in replacement for Windows opengl32.dll, so the import lib cannot be far away, and any bug is theirs. FTR mesa def template is here:
https://gitlab.freedesktop.org/mesa/mesa/-/blob/main/src/gallium/targets/libgl-gdi/opengl32.def.in

**Comment by dan-shaw** [comment]
I don't think we can remove the SDK libs from `opengl`, as it would break msbuild automatic linkage (#99). We may need to skip testing `mesa` in CI.

**Comment by Neumann-A** [comment]
> We may need to skip testing mesa in CI.

no

> as it would break msbuild automatic linkage

If you need opengl in a windows project you should not really use vcpkg auto linking.....

**Comment by dg0yt** [comment]
Can't msbuild use the Windows SDK without copying the needed import libs or DLLs? 
And there is no single port in vcpkg which breaks when they are not present in `installed`?

**Comment by ras0219-msft** [comment]
Some libraries from the Windows SDK are automatically linked by the compiler (`kernel32.lib`), and some are not (`opengl32.lib`). vcpkg intends to provide:
1. A complete description of the direct dependencies of projects
2. (within a single build graph) Unique ownership for each library and symbol
3. Automatic availability of headers and linking of needed libraries for MSBuild

With those points:

> Can't msbuild use the Windows SDK without copying the needed import libs or DLLs?

This could be possible by implementing an opengl32.lib-specific hack, but does not solve ownership -- see below. The file conflicts are not the root cause of the problem, they are a symptom of the true cause: unclear ownership over `opengl32.lib` and `GL/gl.h`.

> And there is no single port in vcpkg which breaks when they are not present in installed?

OSS libraries (so far) are either 1) CMake, which needs to explicitly find_library & link anyway or (rarely) 2) MSBuild projects independent of vcpkg, which will directly list opengl32.lib as an additional linker input file. It does not surprise me that we do not (yet) have a port that requires the automatic linkage. We _know_ we have users that use it -- they've filed issues early on in the life of vcpkg that were fixed by this. 

> We may need to skip testing mesa in CI.

We do not need to skip testing mesa in CI. However, our `opengl` must route to the unique owner of `opengl` -- either the Windows SDK, or system packages, or mesa. This is another case of alternative implementations of a common interface.

This means we need to move compilation and deployment of opengl32.lib / .dll from Mesa into an off-by-default feature so that, across the tree, the Windows SDK is the default owner of opengl. We may need to make similar adjustments to the `opengl-registry` port.

Then, the same as with `openssl`, `blas`, or any of the other interfaces that have multiple providers, users can replace the WinSDK default implementation of `opengl` via and overlay that instead depends on `mesa[opengl]`. This is the only way to ensure that _every_ port consuming opengl will be correctly sequenced after mesa/opengl-registry and use the deterministic and correct libs and headers.

**Comment by Neumann-A** [comment]
@ras0219-msft:

> It does not surprise me that we do not (yet) have a port that requires the automatic linkage.

That probably will never happen if #26370 gets merged since that will always explicitly link libraries in msbuild. As mentioned countless times the current MSBuild integration is flawed by design. Another approach will be needed in the future to avoid further issues with it.

> unclear ownership over opengl32.lib and GL/gl.h.

ownership of opengl32.lib isn't required (more later). Ownership of `GL/gl.h` should be `opengl-registry` because it is the correct upstream. Also I want to point out this is only a Windows issue. On Unix everything needs to move through `opengl-registry` (headers) and system `libglvnd` (dispatch) and then `mesa` (implementation) and there is no selection. 

> This is another case of alternative implementations of a common interface.
> This means we need to move compilation and deployment of opengl32.lib / .dll from Mesa into an off-by-default feature so that,

It is another case but it is different from `ssl/blas/lapack` in that regard that `opengl32` can only be dynamically linked and you generally **DON'T** want to deploy the windows version because that will prohibit your graphics driver from taking over and you will be stuck in an ancient opengl version (1.1) probably not able to run your application. If you want to deploy SW rendering you are going to want to deploy the mesa llvmpipe version.
Again since it is always a dynamic libraries `opengl32.lib` is just a way to tell the linker: I have these symbols please don't error on me. It really doesn't matter if it is even in the installed tree. It could be generated by a def file without being associated with either the WindowsSDK or Mesa as long as the list of symbols is correct. As such ownership is really not the issue here. 
Furthermore opengl functions are always loaded dynamically https://www.khronos.org/opengl/wiki/Load_OpenGL_Functions
by testing the functionality of the driver which makes the real owner of the symbols the driver itself. This makes this case clearly different from the `ssl/blas/lapack` ports.

As such I am still voting for leaving the WindowsSDK out of vcpkg. (Since the alternative would be to have the whole SDK in vcpkg). If you want to have working MSBuild autolinking with `opengl32` just add a `#pragma comment(lib, "opengl32")` into the `GL/gl.h` header for MSVC.

**Comment by dg0yt** [comment]
I reverted the last commit. SDK headers and import lib are installed as before.

**Comment by dg0yt** [comment]
@ras0219-msft @BillyONeal This should be good enough now to move forward from the baseline regression.

**Comment by dg0yt** [comment]
Ping @ras0219-msft @BillyONeal.

**Review by vicroms — COMMENTED** [review]
This LGTM as a stopgap solution.

**Pull request merged**